commit 9197c08532db1c49045f2b79878d62d30ce8190a (refs/changes/40/5040/5)
Author: Brian Bennett <brian.bennett@joyent.com>
Date:   2018-11-29T12:04:23-08:00 (10 months ago)
    
    OS-7361 SmartOS initial set up password prompt mangles some patterns
    Reviewed by: Mike Gerdts <mike.gerdts@joyent.com>
    Reviewed by: Jorge Schrauwen <sjorge@blackdot.be>
    Approved by: Mike Gerdts <mike.gerdts@joyent.com>

diff --git a/overlay/generic/smartdc/lib/smartos_prompt_config.sh b/overlay/generic/smartdc/lib/smartos_prompt_config.sh
old mode 100644
new mode 100755
index 8674ee8c..39a27121
--- a/overlay/generic/smartdc/lib/smartos_prompt_config.sh
+++ b/overlay/generic/smartdc/lib/smartos_prompt_config.sh
@@ -524,13 +524,19 @@ promptnic()
     val=$mac_addr
 }
 
+readpw()
+{
+	IFS='' read -r -s val
+	printf '%s' "$val"
+}
+
 promptpw()
 {
 	def="$3"
 	key="$4"
 
 	if [[ -n ${key} ]]; then
-		preset_val=$(getanswer "${key}")
+		preset_val="$(getanswer "${key}")"
 	fi
 
 	trap "" SIGINT
@@ -538,16 +544,14 @@ promptpw()
 		val=""
 		while [ -z "$val" ]; do
 			if [[ -n ${preset_val} ]]; then
-				val=${preset_val}
+				val="${preset_val}"
 			else
 				if [ -z "$def" ]; then
 					printf "%s: " "$1"
 				else
 					printf "%s [enter to keep existing]: " "$1"
 				fi
-				stty -echo
-				read val
-				stty echo
+				val="$(readpw)"
 				echo
 			fi
 			if [ -n "$val" ]; then
@@ -561,7 +565,7 @@ promptpw()
 				fi
 			else
 				if [ -n "$def" ]; then
-					val=$def
+					val="$def"
 					return
 				else
 					echo "A value must be provided."
@@ -572,12 +576,10 @@ promptpw()
 		cval=""
 		while [ -z "$cval" ]; do
 			if [[ -n ${preset_val} ]]; then
-				cval=${preset_val}
+				cval="${preset_val}"
 			else
 				printf "%s: " "Confirm password"
-				stty -echo
-				read cval
-				stty echo
+				cval="$(readpw)"
 				echo
 			fi
 			[ -n "$cval" ] && break
